File: 백준/Silver/1874. 스택 수열/스택 수열.py
import sys
input = sys.stdin.readline

n = int(input())
arr = list(int(input().strip()) for _ in range(n))

stack = []
answer = []
i = 1
index = 0
while True:
    if not stack:
        # 모든 수열을 봤을 때 성공
        if index == n:
            for j in answer:
                print(j)
            break
        # 정수가 n을 초과했을 때 실패
        if i > n:
            print('NO')
            break
        
        # 수열을 덜 봤을 때 정수 넣어주기
        stack.append(i)
        answer.append('+')
        i += 1
        continue

    if arr[index] == stack[-1]:
        stack.pop()
        answer.append('-')
        index += 1
    elif arr[index] < stack[-1]:
        print('NO')
        break
    else:
        stack.append(i)
        answer.append('+')
        i += 1

File: 백준/Silver/1874. 스택 수열/스택 수열.js
const input = require("fs")
  .readFileSync(
    process.platform === "linux" ? "/dev/stdin" : __dirname + "/input.txt"
  )
  .toString()
  .split("\n")
  .map(Number);

const n = input.shift();
const arr = input;

let cur_num = 1;
const stack = [];
const answer = [];

for (let i = 0; i < n; i++) {
  while (true) {
    if (stack.length && stack[stack.length - 1] === arr[i]) {
      stack.pop();
      answer.push("-");
      break;
    }

    if (cur_num === n + 1) {
      console.log("NO");
      return;
    }

    stack.push(cur_num);
    cur_num++;
    answer.push("+");
  }
}

console.log(answer.join("\n"));
